### DISCUSSION TITLE
Lidar ranges not available after API restart

### DISCUSSION THREAD

**Discussion by chrikatz** [cyberbotics/webots#4154] [Q&A]
Hello,

I added a vehicle to my world and placed a rotating lidar in the sensorSlot and set the controller to "extern". After that I use the Python API commands: 
lidar.enable(1)
lidar.enablePointCloud()
ranges = lidar.getLayerRangeImage(0)

Everything is working fine when the simulation is started the first time. After my python code terminates, the ranges result to None in the restarts. 

The only thing that fixes the problem is to reset the simulation (Ctrl+Shift+T), before starting the python code.
But my intent is to restart my code without any changes in Webots.

**Troubleshooting:**
I tried to disable the device before closing, using
self.lidar.disable()
self.lidar.disablePointCloud()

and I tried to reset the simulation using the simulationReset() command. This only resets the time and does not fix my problem.

**Screenshots**
![image](https://user-images.githubusercontent.com/98445961/151111910-75ae02f6-0195-4eb8-bf9f-4b1b7f8f9867.png)

**System**
Ubuntu 18.04
NVIDIA Corporation TU106GLM [Quadro RTX 3000 Mobile / Max-Q] (rev a1)

**Comment by omichel** [comment]
> After my python code terminates, the ranges result to None in the restarts.

I am not sure what you mean here. Why would your python code terminate? Robot controllers in Webots are supposed to run an endless loop, they never terminate by themselves. They will terminate only when Webots orders them to do so by returning `-1` to the `Robot.step()` method. After this point, your controller shouldn't read any sensor information as the simulation is over. Instead it should clean-up and quit.

**Reply by ShuffleWire** [reply]
It's caused by this bug : https://github.com/cyberbotics/webots/issues/5298
What the OP means, is that when he shut down is controller, then restart it, webots don't provide lidar data anymore, even if everything else work just fine.
He is not in the inverse case, where the simulation would have been restarted, and the controller try to read data from it, even if there are none, obviously.

**Comment by stefaniapedrazzi** [comment]
I'm trying to reproduce the error described here.
However, in all the tests I did if the python controller terminates by itself and it doesn't crash or it is killed then when restarting it the Lidar ranges are still correctly available.

@chrikatz, could you please confirm that your controller is terminating correctly?
If yes, could you provide a complete example where the issue is reproducible?

**Reply by ShuffleWire** [reply]
I just found out something ! If, in the C api, the wb_robot_cleanup() is called before closing controller, then the reconnect goes ok !
what happens in your testing should be that the Python API execute wb_robot_cleanup() whenever the code crash or terminate correctly.
Hence, no issue should be seen in Python.
So in that case, I don't understand as well why the OP got issue, but there could be some strange behavior at play, here...
